testing/asserts.ts

```diff
@@ -864,3 +866,58 @@ export function unimplemented(msg?: string): never {
 export function unreachable(): never {
   throw new AssertionError("unreachable");
 }
+
+let snapshotFile: Record<string, string> | undefined = undefined;
+const updatedSnapshotFile: Record<string, unknown> = {};
+const snapshotMap: Record<string, number> = {};
+
+export async function assertSnapshot(context: Deno.TestContext, actual: unknown) {
+  const name = getName(context);
+  const count = getCount();
+  const testName = `${name} #${count}`;
+  const isUpdate = Deno.args.includes('--update');
+  if (!snapshotFile) {
+    const snapshotPath = getSnapshotPath();
+    await ensureFile(snapshotPath);
+    const {snapshot} = await import(snapshotPath);
+    snapshotFile = snapshot;
+  }
+  const snapshot = snapshotFile?.[testName] ?? '';
+  if (isUpdate) {
+    let isEqual = true;
+    try {
+      assertEquals(_format(actual), snapshot.slice(1, -1));
+    } catch {
+      isEqual = false;
+    }
+    if (!isEqual) console.info("Snapshot updated", testName);
+    updatedSnapshotFile[testName] = actual;
+    globalThis.onunload = writeSnapshotFileSync;
+  } else {
+    assertEquals(_format(actual), snapshot.slice(1, -1));
+  }
+  function getName(context: Deno.TestContext): string {
+    if (context.parent) return `${getName(context.parent)} > ${context.name}`;
+    return context.name;
+  }
+  function getCount() {
+    const count = snapshotMap?.[name] ? snapshotMap[name] : 1;
+    snapshotMap[name] = count + 1;
+    return count;
+  }
+  function getSnapshotPath() {
+    const testFile = fromFileUrl(context.origin);
+    const parts = parse(testFile);
+    return `${join(parts.dir, parts.name)}.snap`;
+  }
+  function writeSnapshotFileSync() {
+    const snapshotPath = getSnapshotPath();
+    ensureFileSync(snapshotPath);
+    const buf = ['export const snapshot = {};\n'];
+    for (const [key, value] of Object.entries(updatedSnapshotFile)) {
+      buf.push(`\nsnapshot[\`${key}\`] = \`\n${_format(value)}\n\`;\n`);
```

Comment:
This doesn't need to be done now, but Jest only adds the extra `\n`s when the formatted result contains a new line.

e.g.

```ts
// test file (jest)
expect("single line").toMatchSnapshot();

// snapshot file
exports[...] = `"single line"`;

// test file (jest)
expect("two\nlines").toMatchSnapshot();

// snapshot file
exports[...] = `
"two
lines"
`;
```

It would be nice if we could mimic this behaviour to make migration of tests easier. Again though, this can be a future improvement if you don't want to add it here.

Reply:
Done